src/librbd/librbd.cc

```diff
@@ -2674,7 +2674,7 @@ namespace librbd {
                ictx->read_only, ofs, len, cmp_bl.length() < len ? NULL : cmp_bl.c_str(),
                bl.length() < len ? NULL : bl.c_str(), op_flags);
 
-    if (bl.length() < len) {
+    if (bl.length() < len || cmp_bl.length() > len) {
       tracepoint(librbd, write_exit, -EINVAL);
```

Comment:
Unrelated, but please fold this in:
```suggestion
      tracepoint(librbd, compare_and_write_exit, -EINVAL);
```